src/budget/CategoryRow.js

```diff
@@ -1,15 +1,129 @@
 import React from 'react';
+import { PropTypes } from 'prop-types';
+import { useSelector } from 'react-redux';
+import Button from 'react-bootstrap/Button';
+import ProgressBar from 'react-bootstrap/ProgressBar';
 
 /**
  * Shows the category title, a progress bar, how much was spent out of the limit
  * and a button to view more details
+ * expects to go in a thing with a variable of whatever
  *
  * @component
  * @example
  *
+ * PUT AN EXAMPLE HERE
  */
-export default function CategoryRow() {
+
+// Need to do:
+// send data in details link
+//  ?what happens on the details page?
+//   <Route path="budget/:categoryId" element={<CategoryDetailsPage />} />
+// ?how should I be getting the time period?
+//   Hard code for now until there's a means to get the desired period.
+// test edge cases
+// edge cases:
+// underspent (return): $0 out of $50 (crossed out) $90 with empty bar
+// overspent: $120 out of $100 with full bar
+
+function CategoryRow({ category }) {
+  const startDate = new Date('06-01-2021');
+  const endDate = new Date('06-30-2021');
+  // check out memoization
+  const amountSpent = useSelector((state) => state.transactions.entities.filter(
+    (transaction) => transaction.categoryId === category.id
+      && transaction.date >= startDate && transaction.date <= endDate,
+  ).reduce((accumulator, currentValue) => accumulator + currentValue.amount, 0).toFixed(2)) * -1;
+
+  const varianceToPercentage = (amountSpent / category.target) * 100;
+
+  let barColor = '';
+  let message = (
+    <span>
+      $
+      {amountSpent.toFixed(2)}
+      &nbsp;spent out of&nbsp;
+      $
+      {category.target.toFixed(2)}
+    </span>
+  );
+
+  if (varianceToPercentage < 0) {
+    message = (
+      <span>
+        $0 spent out of
+        <strike>
+          $
+          {category.target.toFixed(2)}
+        </strike>
+        $
+        (
+        {category.target.toFixed(2)}
+        +
+        {amountSpent.toFixed(2)}
+        )
+      </span>
+    );
+  } else if (varianceToPercentage <= 50 && varianceToPercentage >= 0) {
+    barColor = 'success';
+  } else if (varianceToPercentage > 50 && varianceToPercentage <= 75) {
+    barColor = 'warning';
+  } else if (varianceToPercentage > 75) {
+    barColor = 'danger';
+  } else if (varianceToPercentage > 100) {
+    message = (
+      <span>
+        $
+        {amountSpent.toFixed(2)}
+        &nbsp;spent out of&nbsp;
+        $
+        {category.target.toFixed(2)}
+      </span>
+    );
+    barColor = 'danger'; // add stripe
+  }
+
   return (
-    <div>CategoryRow</div>
+    <div className="container mb-2 px-2 py-2 border border-info rounded">
+      <div className="row px-2">
+        <div className="col col-3 px-2 fw-bold">
```

Comment:
We don't need both the col and the col-3 class. They both apply at the XS size and up, and the col-3 class will overwrite the col class, so we can leave the col class off.

Reply:
Done